AMReX_FabConv.cpp warning writing 1 byte into a region of size 0 [-Wstringop-overflow=]

Dear AMReX developers,
cc: @ax3l 
I am a `WarpX` user in NERSC.
I observed the following warning message when I was building `WarpX` in Perlmutter.

```
In function 'void amrex::_pd_btrvout(char*, amrex::Long, amrex::Long)',
    inlined from 'void amrex::_pd_insert_field(amrex::Long, int, char*, int, int, int)' at /build2204-hdf5/_deps/fetchedamrex-src/Src/Base/AMReX_FabConv.cpp:462:12,
    inlined from 'void amrex::PD_fconvert(void*, const void*, amrex::Long, int, const Long*, const int*, const Long*, const int*, int, int, int)' at /build2204-hdf5/_deps/fetchedamrex-src/Src/Base/AMReX_FabConv.cpp:743:17:
/build2204-hdf5/_deps/fetchedamrex-src/Src/Base/AMReX_FabConv.cpp:393:7: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  393 |             *p2 = tmp;
      |       ^
/build2204-hdf5/_deps/fetchedamrex-src/Src/Base/AMReX_FabConv.cpp: In function 'void amrex::PD_fconvert(void*, const void*, amrex::Long, int, const Long*, const int*, const Long*, const int*, int, int, int)':
/build2204-hdf5/_deps/fetchedamrex-src/Src/Base/AMReX_FabConv.cpp:612:6: note: at offset [9, 1073741822] into destination object 'in_long' of size 8
  612 | PD_fconvert (void*       out,
```

Here's the list of modules that I use in Perlmutter.
```
Currently Loaded Modules:
  1) craype-x86-milan                      10) craype/2.7.15
  2) libfabric/1.11.0.4.114                11) cray-dsmml/0.2.2
  3) craype-network-ofi                    12) cray-mpich/8.1.15
  4) perftools-base/22.04.0                13) cray-libsci/21.08.1.2
  5) xpmem/2.3.2-2.2_7.5__g93dd7ee.shasta  14) PrgEnv-gnu/8.3.3
  6) xalt/2.10.2                           15) Nsight-Compute/2022.1.1
  7) darshan/3.3.1                         16) Nsight-Systems/2022.2.1
  8) cmake/3.22.0                          17) cudatoolkit/11.5
  9) gcc/11.2.0                            18) cray-hdf5-parallel/1.12.1.1

```
Thank you.

--- a/Src/Base/AMReX_FabConv.cpp
+++ b/Src/Base/AMReX_FabConv.cpp
@@ -375,15 +375,14 @@ _pd_extract_field (char const* in,
 // Byte reverse nitems words.  Each word is nb bytes Long where nb is even.
 //
 
+template <int NB>
 static
 void
-_pd_btrvout (char* out,
-             Long  nb,
-             Long  nitems)
+_pd_btrvout (char* out, Long  nitems)
 {
-    for (Long jl = 0, nbo2 = nb >> 1; jl < nbo2; jl++)
+    for (int jl = 0, nbo2 = NB >> 1; jl < nbo2; jl++)
     {
-        Long jh  = nb - jl - 1;
+        int jh  = NB - jl - 1;
         char* p1 = out + jh;
         char* p2 = out + jl;
         for (Long i = 0L; i < nitems; i++)
@@ -391,8 +390,8 @@ _pd_btrvout (char* out,
             char tmp = *p1;
             *p1 = *p2;
             *p2 = tmp;
-            p1 += nb;
-            p2 += nb;
+            p1 += NB;
+            p2 += NB;
         }
     }
 }
@@ -458,8 +457,13 @@ _pd_insert_field (Long  in_long,
     //
     // Reorder the bytes appropriately.
     //
-    if (l_order == REVERSE_ORDER)
-        _pd_btrvout(in, l_bytes, 1L);
+    if (l_order == REVERSE_ORDER) {
+        if (l_bytes == 4) {
+            _pd_btrvout<4>(in, 1L);
+        } else { // It's either 4 or 8.  There is an assertion in PD_fconvert.
+            _pd_btrvout<8>(in, 1L);
+        }
+    }
     //
     // Copy the remaining aligned bytes over.
     //
@@ -622,6 +626,9 @@ PD_fconvert (void*       out,
              int         onescmp)
 {
 //    BL_PROFILE("PD_fconvert");
+
+    AMREX_ASSERT(l_bytes == 4 || l_bytes == 8); // Otherwise, we need to update _pd_btrvout
+
     Long i, expn, expn_max, hexpn, mant, DeltaBias, hmbo, hmbi;
     int nbits, inbytes, outbytes, sign;
     int indxin, indxout, inrem, outrem, dindx;

